### PR TITLE
Clear charts when clearing plotters

### PR DESCRIPTION
**Pull request by adeak**
[This Stack Overflow question](https://stackoverflow.com/questions/72985319/pyvista-why-does-the-callback-only-draw-the-3d-entities-and-not-the-2d-one-on-t/) has an excellent minimal example wherein calling `plotter.clear()` doesn't clear the underlying `Charts` object, which prevents replotting from succeeding.

It's still not clear to me what the exact mechanics are that prevent replotting, but it's clear (no pun intended) that
1. we _should_ clear charts on `plotter.clear()`, and
2. doing so fixes the bug.

For others as unfamiliar with these execution paths as I am:
1. `pyvistaqt.QtInteractor` inherits its `.clear()` method from `BasePlotter`
2. `BasePlotter` delegates it to `plotter.renderers.clear()`
3. `Renderers` delegates it to `plotter.renderer.clear()`
4. `Renderer.clear()` is the final source: https://github.com/pyvista/pyvista/blob/8f47b9e479737dcbe1de8bba1eb12c6688b8b9ce/pyvista/plotting/renderer.py#L1862-L1876

Calling `self._charts.deep_clean()` in there will make sure to clear the internal `Charts` object (after creating one from scratch if necessary; this is the cost of not messing with `__charts` directly), then removes the `Charts`' actor from the renderer.

---

A few things stood out to me while I was investigating this:
- There's no `Charts.clear()`, perhaps `deep_clean()` should be called that instead?
- `Charts` doesn't use the renderer's `add_actor()`/`remove_actor()` methods, which doesn't seem nice. E.g. `plotter.renderer._actors` has no knowledge of any chart-related stuff this way. Unless there's a good reason for this, we should fix this (potentially in a different PR; this bugfix would be nice to have in the next patch release).

### CODE REVIEW

**Comment by adeak** [comment]
Welp, I broke the labeler action again :cold_sweat:

File: pyvista/plotting/renderer.py [hunk 1]
```diff
@@ -1868,6 +1868,7 @@ def clear(self):
                 except KeyError:
                     pass
 
+        self._charts.deep_clean()
```
Comment (adeak) [inline pyvista/plotting/renderer.py:1871]: Accessing `self._charts` in `clear()` (even without the `deep_clean()`) leads to GC errors in `test_plotting.py`, due to some `dataset` `vtkWeakReference` not being collecting. Trying to figure this out...

Comment (adeak) [reply pyvista/plotting/renderer.py]: I guess the issue was related to creating a new `Charts` instance via the `self._charts` property when it was `None` earlier. Adding a less elegant explicit check for `self.__charts is not None` seems to make the GC errors go away, while still fixing the original issue. I'm open to alternate suggestions though.

File: pyvista/plotting/charts.py [hunk 2]
```diff
@@ -4458,3 +4458,7 @@ def __iter__(self):
         """Return an iterable of charts."""
         for chart in self._charts:
             yield chart
+
+    def __del__(self):
```
Comment (akaszynski) [inline pyvista/plotting/charts.py:4462]:
```suggestion
    def __del__(self):  # pragma: no cover
```

**Review by akaszynski — APPROVED** [review]
LGTM. Thanks for looking into this @adeak!

File: pyvista/plotting/charts.py (hunk 2 above)
Comment (adeak) [reply pyvista/plotting/charts.py]: This actually led me down a rabbit hole that I'll open a PR about soon, but I'm still investigating (this is why I haven't yet looked at the other PRs as planned, sorry). I should've pushed this for the time being, thanks for adding it.

**Review by tkoyama010 — APPROVED** [review]
LGTM

**Pull request merged**